### PR TITLE
Use RELATIVE path sensitivity for JavaCompile sources [WIP]

### PR DESCRIPTION
**Pull request by lptr**
Previously `JavaCompile` used to ignore the path of its source files, only caring about the file names.
While the `NAME_ONLY` sensitivity used in the past was more correct in the sense that `javac` ignores its input files' paths, the current `RELATIVE` sensitivity has some advantages.
For one, it is more in line with other compilation tasks like `GroovyCompile` and `CppCompile`.
In the vast majority of cases Java classes should be organized according to their package hierarchy, which means `RELATIVE` behavior is expected.
Due to some recent optimizations in task input handling using `RELATIVE` sensitivity also means Gradle can be faster in detecting changes and accessing cached results.

### CODE REVIEW

File: subprojects/docs/src/docs/release/notes.md [hunk 1]
```diff
@@ -291,6 +291,14 @@ The `IdeaModule` Tooling API model element contains methods to retrieve resource
 In previous Gradle versions the `source` filed in `SourceTask` was accessible from subclasses.
 This is not the case anymore as the `source` filed is now declared as `private`.
 
+### `JavaCompile` task path sensitivity change
+
+Previously `JavaCompile` used to ignore the path of its source files, only caring about the file names.
+While the `NAME_ONLY` sensitivity used in the past was more correct in the sense that `javac` ignores its input files' paths, the current `RELATIVE` sensitivity has some advantages.
+For one, it is more in line with other compilation tasks like `GroovyCompile` and `CppCompile`.
+In the vast majority of cases Java classes should be organized according to their package hierarchy, which means `RELATIVE` behavior is expected.
+Due to some recent optimizations in task input handling using `RELATIVE` sensitivity also means Gradle can be faster in detecting changes and accessing cached results.
```
Comment (wolfs) [inline subprojects/docs/src/docs/release/notes.md]: 👍

File: subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/JavaCompile.java [hunk 2]
```diff
@@ -75,7 +75,7 @@ public JavaCompile() {
      * {@inheritDoc}
      */
     @Override
-    @PathSensitive(PathSensitivity.NAME_ONLY)
+    @PathSensitive(PathSensitivity.RELATIVE)
```
Comment (wolfs) [inline subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/JavaCompile.java]: I was surprised that only one test failed. I thought we still had some relocation tests which moved some classes around in the source directory.

**Comment by lptr** [comment]
This needs some more work to actually add the optimization. Without it we have a performance regression.

**Comment by lptr** [comment]
We decided not to pursue this for 5.0.

**Pull request closed without merging**